### PR TITLE
[Inference API] Add support for third party models to Semantic Text

### PR DESCRIPTION
**Pull request by maxhniebergall**
This change adds support for third party models to semantic text, and includes some refactoring.

One of the major refactors was to change the names of the results objects. Previously, many of the results objects had the same names as some of the objects from `ml.trainedmodels` which resulted in confusion and some cases of the wrong results class being used. Now, those objects start with `Inference` and contain the type of result `Byte` or `Float`.

I tested this change with my big data test python notebook in cloud with openAI.

### CODE REVIEW

File: x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldTests.java [hunk 1]
```diff
@@ -132,33 +138,37 @@ public void testModelSettingsValidation() {
...
-    public static ChunkedTextEmbeddingResults randomTextEmbeddings(Model model, List<String> inputs) {
-        List<org.elasticsearch.xpack.core.ml.inference.results.ChunkedTextEmbeddingResults.EmbeddingChunk> chunks = new ArrayList<>();
+    public static InferenceChunkedTextEmbeddingFloatResults randomInferenceChunkedTextEmbeddingFloatResults(
+        Model model,
+        List<String> inputs
+    ) throws IOException {
+        List<InferenceChunkedTextEmbeddingFloatResults.InferenceFloatEmbeddingChunk> chunks = new ArrayList<>();
         for (String input : inputs) {
-            double[] values = new double[model.getServiceSettings().dimensions()];
+            float[] values = new float[model.getServiceSettings().dimensions()];
```
Comment (maxhniebergall) [inline x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldTests.java:147]: switched from double to float as the results classes have float arrays under the hood

File: server/src/main/java/org/elasticsearch/inference/ChunkedInferenceServiceResults.java [hunk 2]
```diff
@@ -8,6 +8,19 @@
...
 package org.elasticsearch.inference;
 
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.xcontent.XContent;
+
+import java.util.Iterator;
+
 public interface ChunkedInferenceServiceResults extends InferenceServiceResults {
 
+    Iterator<Chunk> chunksAsMatchedTextAndByteReference(XContent xcontent);
```
Comment (maxhniebergall) [inline server/src/main/java/org/elasticsearch/inference/ChunkedInferenceServiceResults.java:18]: Needed to support semantic text

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/inference/results/ErrorChunkedInferenceResults.java [hunk 3]
```diff
@@ -94,4 +97,9 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
     public String getWriteableName() {
         return NAME;
     }
+
+    @Override
+    public Iterator<Chunk> chunksAsMatchedTextAndByteReference(XContent xcontent) {
+        return List.of(exception).stream().map(e -> new Chunk(e.getMessage(), (BytesReference) null)).iterator();
```
Comment (maxhniebergall) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/inference/results/ErrorChunkedInferenceResults.java:103]: it should be ok to return a null bytes reference, right?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/inference/results/InferenceChunkedSparseEmbeddingResults.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.xpack.core.inference.results.TextEmbeddingUtils.validateInputSizeAgainstEmbeddings;
+
+public class InferenceChunkedSparseEmbeddingResults implements ChunkedInferenceServiceResults {
```
Comment (maxhniebergall) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/inference/results/InferenceChunkedSparseEmbeddingResults.java:31]: this file was actually just renamed

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/inference/results/InferenceChunkedTextEmbeddingByteResults.java [hunk 5]
```diff
@@ -156,4 +161,22 @@ public int hashCode() {
...
+            b.startArray();
+            for (byte v : value) {
+                b.value(v);
+            }
+            b.endArray();
+            return BytesReference.bytes(b);
+        } catch (IOException exc) {
+            throw new RuntimeException(exc);
+        }
+    }
```
Comment (maxhniebergall) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/inference/results/InferenceChunkedTextEmbeddingByteResults.java:181]: main addition for semantic text in this class

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/inference/results/InferenceChunkedSparseEmbeddingResults.java [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+            b.startObject();
+            for (var weightedToken : tokens) {
+                weightedToken.toXContent(b, ToXContent.EMPTY_PARAMS);
+            }
+            b.endObject();
+            return BytesReference.bytes(b);
+        } catch (IOException exc) {
+            throw new RuntimeException(exc);
+        }
+    }
```
Comment (maxhniebergall) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/inference/results/InferenceChunkedSparseEmbeddingResults.java:155]: this is the new code

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/common/EmbeddingRequestChunker.java [hunk 7]
```diff
@@ -160,7 +160,7 @@ private class DebatchingListener implements ActionListener<InferenceServiceResul
 
         @Override
         public void onResponse(InferenceServiceResults inferenceServiceResults) {
-            if (inferenceServiceResults instanceof TextEmbeddingResults textEmbeddingResults) { // TODO byte embeddings
+            if (inferenceServiceResults instanceof InferenceTextEmbeddingFloatResults textEmbeddingResults) { // TODO byte embeddings
```
Comment (davidkyle) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/common/EmbeddingRequestChunker.java:163]: > // TODO byte embeddings

Byte embeddings and sparse embeddings can be done in the follow up

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java [hunk 8]
```diff
@@ -292,60 +286,13 @@ public static List<Chunk> toSemanticTextFieldChunks(
...
-                    "Invalid inference results format for field [{}] with inference id [{}], got {}",
-                    RestStatus.BAD_REQUEST,
-                    field,
-                    inferenceId,
-                    result.getWriteableName()
-                );
+            for (Iterator<ChunkedInferenceServiceResults.Chunk> it = result.chunksAsMatchedTextAndByteReference(contentType.xContent()); it
+                .hasNext();) {
+                var chunkAsByteReference = it.next();
+                chunks.add(new Chunk(chunkAsByteReference.matchedText(), chunkAsByteReference.bytesReference()));
```
Comment (davidkyle) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java:292]: nice

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/inference/results/ErrorChunkedInferenceResults.java (hunk 3 above)
Comment (carlosdelest) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/inference/results/ErrorChunkedInferenceResults.java:103]: Yes - SemanticText is not using this implementation, just checking when it's receiving an instance of the class to mark the inference as failed.

We could as well throw an `UnsupportedOperationException` for this

File: x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/services/azureopenai/AzureOpenAiServiceTests.java [hunk 9]
```diff
@@ -1066,29 +1066,30 @@ public void testInfer_UnauthorisedResponse() throws IOException, URISyntaxExcept
         }
     }
 
+    // TODO
```
Comment (davidkyle) [inline x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/services/azureopenai/AzureOpenAiServiceTests.java:1069]: > // TODO

? looks done to me

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java [hunk 10]
```diff
@@ -292,60 +286,13 @@ public static List<Chunk> toSemanticTextFieldChunks(
...
-                }
-            } else {
-                throw new ElasticsearchStatusException(
-                    "Invalid inference results format for field [{}] with inference id [{}], got {}",
-                    RestStatus.BAD_REQUEST,
-                    field,
-                    inferenceId,
-                    result.getWriteableName()
-                );
+            for (Iterator<ChunkedInferenceServiceResults.Chunk> it = result.chunksAsMatchedTextAndByteReference(contentType.xContent()); it
```
Comment (carlosdelest) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java:289]: This looks so much better now, and we have support for upcoming classes 💯

**Review by davidkyle — COMMENTED** [review]
Looks great, thanks for the renaming that will make code navigation easier.

CI is green but I'd like to run some more tests on the PR before I approve

**Review by carlosdelest — APPROVED** [review]
Thanks for the changes @maxhniebergall ! Much cleaner now from `SemanticText` 💯

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/inference/results/ErrorChunkedInferenceResults.java (hunk 3 above)
Comment (maxhniebergall) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/inference/results/ErrorChunkedInferenceResults.java:103]: I changed it to use `BytesArray::EMPTY` as I'm not sure what the benefit would be of throwing an exception instead, but I'm open to that option.

**Review by davidkyle — APPROVED** [review]
LGTM

**Comment by maxhniebergall** [comment]
closes https://github.com/elastic/elasticsearch/issues/108921

**Pull request merged**